src/chips/safe_accumulator.rs

```diff
@@ -0,0 +1,211 @@
+use num_bigint::BigUint;
+use std::fmt::Debug;
+
+use super::is_zero::{IsZeroChip, IsZeroConfig};
+use super::utils::{decompose_bigInt_to_ubits, fp_to_big_uint, value_fp_to_big_uint, range_check_vec};
+use halo2_proofs::{circuit::*, halo2curves::pasta::Fp, plonk::*, poly::Rotation};
+
+#[derive(Debug, Clone)]
+pub struct SafeAccumulatorConfig<const MAX_BITS: u8, const ACC_COLS: usize> {
+    pub update_value: Column<Advice>,
+    pub left_most_inv: Column<Advice>,
+    pub add_carries: [Column<Advice>; ACC_COLS],
+    pub accumulate: [Column<Advice>; ACC_COLS],
+    pub instance: Column<Instance>,
+    pub is_zero: IsZeroConfig,
+    pub selector: [Selector; 2],
+}
+
+#[derive(Debug, Clone)]
+pub struct SafeACcumulatorChip<const MAX_BITS: u8, const ACC_COLS: usize> {
+    config: SafeAccumulatorConfig<MAX_BITS, ACC_COLS>,
+}
+
+impl<const MAX_BITS: u8, const ACC_COLS: usize> SafeACcumulatorChip<MAX_BITS, ACC_COLS> {
+    pub fn construct(config: SafeAccumulatorConfig<MAX_BITS, ACC_COLS>) -> Self {
+        Self { config }
+    }
+
+    pub fn configure(
+        meta: &mut ConstraintSystem<Fp>,
+        update_value: Column<Advice>,
+        left_most_inv: Column<Advice>,
+        add_carries: [Column<Advice>; ACC_COLS],
```

Comment:
I didn't fully understand the role of `add_carries` here

Reply:
these might help you to understand it. it's for checking constraints.

https://github.com/summa-dev/halo2-experiments/pull/5#discussion_r1182350950

https://github.com/summa-dev/halo2-experiments/pull/5#discussion_r1182444667